File: squire-backend/api/routes/events.js
// User Events Routes
const express = require('express');
const { createClient } = require('@supabase/supabase-js');
const router = express.Router();

const supabase = createClient(
  process.env.SUPABASE_URL,
  process.env.SUPABASE_ANON_KEY
);

// POST /api/events - Create new user event
router.post('/', async (req, res) => {
  try {
    const {
      user_id,
      event_type,
      event_data,
      importance_score,
      tags = [],
      session_id,
      related_suggestion_id
    } = req.body;

    const { data, error } = await supabase
      .rpc('add_user_event', {
        p_user_id: user_id,
        p_event_type: event_type,
        p_event_data: event_data,
        p_importance_score: importance_score,
        p_tags: tags,
        p_session_id: session_id,
        p_related_suggestion_id: related_suggestion_id
      });

    if (error) throw error;

    // Get the created event
    const { data: event, error: eventError } = await supabase
      .from('user_events')
      .select('*')
      .eq('id', data)
      .single();

    if (eventError) throw eventError;

    res.status(201).json({ event, event_id: data });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// GET /api/events/user/:userId - Get all events for user
router.get('/user/:userId', async (req, res) => {
  try {
    const { userId } = req.params;
    const {
      event_type,
      min_importance,
      tags,
      session_id,
      limit = 100,
      offset = 0,
      order_by = 'created_at',
      order_direction = 'desc'
    } = req.query;

    let query = supabase
      .from('user_events')
      .select('*')
      .eq('user_id', userId);

    if (event_type) {
      query = query.eq('event_type', event_type);
    }

    if (min_importance) {
      query = query.gte('importance_score', parseFloat(min_importance));
    }

    if (tags) {
      const tagArray = tags.split(',');
      query = query.overlaps('tags', tagArray);
    }

    if (session_id) {
      query = query.eq('session_id', session_id);
    }

    query = query
      .order(order_by, { ascending: order_direction === 'asc' })
      .range(offset, offset + limit - 1);

    const { data, error } = await query;

    if (error) throw error;

    res.json({ events: data || [] });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// GET /api/events/:id - Get specific event
router.get('/:id', async (req, res) => {
  try {
    const { id } = req.params;

    const { data, error } = await supabase
      .from('user_events')
      .select('*')
      .eq('id', id)
      .single();

    if (error) throw error;
    if (!data) return res.status(404).json({ error: 'Event not found' });

    res.json({ event: data });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// PUT /api/events/:id - Update event
router.put('/:id', async (req, res) => {
  try {
    const { id } = req.params;
    const {
      event_data,
      importance_score,
      tags,
      metadata
    } = req.body;

    const updateData = {};
    if (event_data !== undefined) updateData.event_data = event_data;
    if (importance_score !== undefined) updateData.importance_score = importance_score;
    if (tags !== undefined) updateData.tags = tags;
    if (metadata !== undefined) updateData.metadata = metadata;

    const { data, error } = await supabase
      .from('user_events')
      .update(updateData)
      .eq('id', id)
      .select()
      .single();

    if (error) throw error;
    if (!data) return res.status(404).json({ error: 'Event not found' });

    res.json({ event: data });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// DELETE /api/events/:id - Delete event
router.delete('/:id', async (req, res) => {
  try {
    const { id } = req.params;

    const { data, error } = await supabase
      .from('user_events')
      .delete()
      .eq('id', id)
      .select()
      .single();

    if (error) throw error;
    if (!data) return res.status(404).json({ error: 'Event not found' });

    res.json({ success: true, message: 'Event deleted successfully' });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// POST /api/events/bulk - Create multiple events
router.post('/bulk', async (req, res) => {
  try {
    const { events } = req.body; // Array of event objects

    const results = [];
    for (const event of events) {
      try {
        const { data, error } = await supabase
          .rpc('add_user_event', {
            p_user_id: event.user_id,
            p_event_type: event.event_type,
            p_event_data: event.event_data,
            p_importance_score: event.importance_score,
            p_tags: event.tags || [],
            p_session_id: event.session_id,
            p_related_suggestion_id: event.related_suggestion_id
          });

        if (error) {
          results.push({ success: false, error: error.message, event });
        } else {
          results.push({ success: true, event_id: data, event });
        }
      } catch (err) {
        results.push({ success: false, error: err.message, event });
      }
    }

    const successCount = results.filter(r => r.success).length;
    res.json({
      results,
      total: events.length,
      successful: successCount,
      failed: events.length - successCount
    });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// GET /api/events/user/:userId/analytics - Get event analytics
router.get('/user/:userId/analytics', async (req, res) => {
  try {
    const { userId } = req.params;
    const { days = 30 } = req.query;

    const { data, error } = await supabase
      .from('user_events')
      .select('*')
      .eq('user_id', userId)
      .gte('created_at', new Date(Date.now() - days * 24 * 60 * 60 * 1000).toISOString());

    if (error) throw error;

    const analytics = {
      total: data.length,
      by_type: data.reduce((acc, e) => {
        acc[e.event_type] = (acc[e.event_type] || 0) + 1;
        return acc;
      }, {}),
      avg_importance: data.length ?
        data.reduce((sum, e) => sum + (e.importance_score || 0), 0) / data.length : 0,
      high_importance_count: data.filter(e => e.importance_score > 0.7).length,
      tag_frequency: data.reduce((acc, e) => {
        (e.tags || []).forEach(tag => {
          acc[tag] = (acc[tag] || 0) + 1;
        });
        return acc;
      }, {}),
      daily_counts: data.reduce((acc, e) => {
        const date = new Date(e.created_at).toISOString().split('T')[0];
        acc[date] = (acc[date] || 0) + 1;
        return acc;
      }, {})
    };

    res.json({ analytics });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// GET /api/events/user/:userId/patterns - Get behavioral patterns
router.get('/user/:userId/patterns', async (req, res) => {
  try {
    const { userId } = req.params;
    const { event_type, days = 30 } = req.query;

    let query = supabase
      .from('user_events')
      .select('*')
      .eq('user_id', userId)
      .gte('created_at', new Date(Date.now() - days * 24 * 60 * 60 * 1000).toISOString())
      .order('created_at', { ascending: true });

    if (event_type) {
      query = query.eq('event_type', event_type);
    }

    const { data, error } = await query;

    if (error) throw error;

    // Analyze patterns
    const patterns = {
      frequency: {
        hourly: new Array(24).fill(0),
        daily: {},
        weekly: new Array(7).fill(0)
      },
      sequences: {},
      trends: {
        increasing: [],
        decreasing: [],
        stable: []
      }
    };

    // Calculate frequency patterns
    data.forEach(event => {
      const date = new Date(event.created_at);
      const hour = date.getHours();
      const day = date.toISOString().split('T')[0];
      const weekday = date.getDay();

      patterns.frequency.hourly[hour]++;
      patterns.frequency.daily[day] = (patterns.frequency.daily[day] || 0) + 1;
      patterns.frequency.weekly[weekday]++;
    });

    // Find common sequences (simplified)
    for (let i = 0; i < data.length - 1; i++) {
      const current = data[i].event_type;
      const next = data[i + 1].event_type;
      const sequence = `${current} → ${next}`;
      patterns.sequences[sequence] = (patterns.sequences[sequence] || 0) + 1;
    }

    res.json({ patterns });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// POST /api/events/process-to-knowledge/:userId - Process events to knowledge graph
router.post('/process-to-knowledge/:userId', async (req, res) => {
  try {
    const { userId } = req.params;
    const { batch_size = 100 } = req.body;

    const { data, error } = await supabase
      .rpc('process_events_to_knowledge', {
        p_user_id: userId,
        p_batch_size: batch_size
      });

    if (error) throw error;

    res.json({
      success: true,
      message: `${data} events processed to knowledge graph`,
      processed_count: data
    });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

module.exports = router;

File: squire-backend/app/routers/events.py
"""
User Events routes
"""
from fastapi import APIRouter, HTTPException, Depends, Query
from typing import Optional, List
from uuid import UUID

from app.core.database import get_supabase, execute_rpc, execute_query, DatabaseError
from app.models.schemas import (
    UserEvent,
    UserEventCreate,
    UserEventUpdate,
    UserEventBulk,
    EventType,
    SuccessResponse
)

router = APIRouter()


@router.post("/", response_model=dict)
async def create_user_event(
    event_data: UserEventCreate,
    supabase=Depends(get_supabase)
):
    """Create new user event"""
    try:
        event_id = await execute_rpc(
            "add_user_event",
            {
                "p_user_id": str(event_data.user_id),
                "p_event_type": event_data.event_type.value,
                "p_event_data": event_data.event_data,
                "p_importance_score": event_data.importance_score,
                "p_tags": event_data.tags,
                "p_session_id": str(event_data.session_id) if event_data.session_id else None,
                "p_related_suggestion_id": str(event_data.related_suggestion_id) if event_data.related_suggestion_id else None
            }
        )

        event = await execute_query(
            table="user_events",
            operation="select",
            filters={"id": str(event_id)},
            single=True
        )

        return {
            "event": UserEvent(**event),
            "event_id": event_id
        }
    except DatabaseError as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/user/{user_id}", response_model=dict)
async def get_user_events(
    user_id: UUID,
    event_type: Optional[EventType] = Query(None),
    min_importance: Optional[float] = Query(None, ge=0.0, le=1.0),
    tags: Optional[str] = Query(None),
    session_id: Optional[UUID] = Query(None),
    limit: int = Query(100, le=1000),
    offset: int = Query(0, ge=0),
    order_by: str = Query("created_at"),
    order_direction: str = Query("desc"),
    supabase=Depends(get_supabase)
):
    """Get all events for user"""
    try:
        filters = {"user_id": str(user_id)}

        if event_type:
            filters["event_type"] = event_type.value
        if session_id:
            filters["session_id"] = str(session_id)

        events = await execute_query(
            table="user_events",
            operation="select",
            filters=filters,
            order_by=order_by,
            ascending=(order_direction == "asc"),
            limit=limit,
            offset=offset
        )

        # Filter by min_importance and tags in Python (would be better in DB)
        if events:
            if min_importance is not None:
                events = [e for e in events if e.get("importance_score", 0) >= min_importance]

            if tags:
                tag_list = tags.split(",")
                events = [e for e in events if any(tag in e.get("tags", []) for tag in tag_list)]

        return {"events": [UserEvent(**event) for event in (events or [])]}
    except DatabaseError as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/{event_id}", response_model=UserEvent)
async def get_event(
    event_id: UUID,
    supabase=Depends(get_supabase)
):
    """Get specific event"""
    try:
        event = await execute_query(
            table="user_events",
            operation="select",
            filters={"id": str(event_id)},
            single=True
        )

        if not event:
            raise HTTPException(status_code=404, detail="Event not found")

        return UserEvent(**event)
    except DatabaseError as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/bulk", response_model=dict)
async def create_events_bulk(
    events_data: UserEventBulk,
    supabase=Depends(get_supabase)
):
    """Create multiple events"""
    try:
        results = []
        for event in events_data.events:
            try:
                event_id = await execute_rpc(
                    "add_user_event",
                    {
                        "p_user_id": str(event.user_id),
                        "p_event_type": event.event_type.value,
                        "p_event_data": event.event_data,
                        "p_importance_score": event.importance_score,
                        "p_tags": event.tags,
                        "p_session_id": str(event.session_id) if event.session_id else None,
                        "p_related_suggestion_id": str(event.related_suggestion_id) if event.related_suggestion_id else None
                    }
                )
                results.append({"success": True, "event_id": event_id, "event": event.dict()})
            except Exception as e:
                results.append({"success": False, "error": str(e), "event": event.dict()})

        successful_count = sum(1 for r in results if r["success"])
        return {
            "results": results,
            "total": len(events_data.events),
            "successful": successful_count,
            "failed": len(events_data.events) - successful_count
        }
    except DatabaseError as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/process-to-knowledge/{user_id}", response_model=dict)
async def process_events_to_knowledge(
    user_id: UUID,
    batch_size: int = Query(100, ge=1, le=1000),
    supabase=Depends(get_supabase)
):
    """Process events to knowledge graph"""
    try:
        count = await execute_rpc(
            "process_events_to_knowledge",
            {
                "p_user_id": str(user_id),
                "p_batch_size": batch_size
            }
        )

        return {
            "success": True,
            "message": f"{count} events processed to knowledge graph",
            "processed_count": count
        }
    except DatabaseError as e:
        raise HTTPException(status_code=500, detail=str(e))